extensions/data-transfer/portability-data-transfer-google/src/main/java/org/dataportabilityproject/datatransfer/google/mail/GoogleMailExporter.java

```diff
@@ -0,0 +1,132 @@
+
+public class GoogleMailExporter implements Exporter<TokensAndUrlAuthData, MailContainerResource> {
+  private static final long PAGE_SIZE = 50;
+  private static final String USER = "me";
+  private volatile Gmail gmail;
+
+  public GoogleMailExporter() {
+    this.gmail = null;
+  }
+
+  @VisibleForTesting
+  GoogleMailExporter(Gmail gmail) {
+    this.gmail = gmail;
+  }
+
+  @Override
+  public ExportResult<MailContainerResource> export(UUID id, TokensAndUrlAuthData authData) {
+    return export(id, authData, new ExportInformation(null, null));
+  }
+
+  @Override
+  public ExportResult<MailContainerResource> export(UUID id,
+      TokensAndUrlAuthData authData, ExportInformation exportInformation) {
+    // Create a new gmail service for the authorized user
+    Gmail gmail = getOrCreateGmail(authData);
+
+    IdOnlyContainerResource resource =
+        (IdOnlyContainerResource) exportInformation.getContainerResource();
+
+    Messages.List request = null;
+    try {
+      request = gmail.users().messages().list(USER).setMaxResults(PAGE_SIZE);
+    } catch (IOException e) {
+      return new ExportResult(
+          ExportResult.ResultType.ERROR, "Error creating request: " + e.getMessage());
+    }
+
+    if (exportInformation.getPaginationData() != null) {
+      request.setPageToken(
+          ((StringPaginationToken) exportInformation.getPaginationData()).getToken());
+    }
+
+    ListMessagesResponse response = null;
+    try {
+      response = request.execute();
+    } catch (IOException e) {
+      return new ExportResult(
+          ExportResult.ResultType.ERROR, "Error exporting messages: " + e.getMessage());
+    }
+
+    List<MailMessageModel> results = new ArrayList<>(response.getMessages().size());
+    // TODO: this is a good indication we need to swap the interface
+    // as we can't store all the mail messagess in memory at once.
+    for (Message listMessage : response.getMessages()) {
+      Message getResponse = null;
+      try {
+        getResponse =
+            gmail.users().messages().get(USER, listMessage.getId()).setFormat("raw").execute();
+      } catch (IOException e) {
+        return new ExportResult(
+            ExportResult.ResultType.ERROR, "Error exporting single message: " + e.getMessage());
+      }
+      // TODO: note this doesn't transfer things like labels
+      results.add(new MailMessageModel(getResponse.getRaw(), getResponse.getLabelIds()));
+    }
+
+    PaginationData newPage = null;
+    ResultType resultType = ResultType.END;
+    if (response.getNextPageToken() != null) {
+      newPage = new StringPaginationToken(response.getNextPageToken());
+      resultType = ResultType.CONTINUE;
+    }
+
+    MailContainerResource mailContainerResource = new MailContainerResource(null, results);
+    return new ExportResult<>(resultType, mailContainerResource, new ContinuationData(newPage));
+  }
+
+  private Gmail getOrCreateGmail(TokensAndUrlAuthData authData) {
```

Comment:
maybe Gmail could be provided by Guice. Not sure if we are still planning injecting stuff for the Google extensions (I think we still do for the cloud extension). Fine to leave for now but might want to look at that later.